File: my-folder/0075-sort-colors/solution.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        red = 0
        white = 0
        blue = len(nums) - 1
        #red pushes all the numbers greater than 0 to the right
        #blue moves ahead after setting all the 2's at the end of the array
        while white <= blue:
            if nums[white] == 0:
                nums[white], nums[red] = nums[red], nums[white]
                red += 1
                white += 1
            # only when the nums[white] is 1 then increment white this can be case where blue and white came to same position and we almost sorted  the array
            elif nums[white] == 1:
                white += 1 
            # when the nums[white] is 2 then we swap with blue decr blue 
            else:
                nums[white], nums[blue] = nums[blue], nums[white]
                blue -= 1
                
# [2,0,2,1,1,0]
#  l r

File: my-folder/0075-sort-colors/solution.java
class Solution {
    public void sortColors(int[] nums) {
        int left = 0;
        int right = nums.length - 1;
        
        int curr = 0;
        while(curr <= right){
            if(nums[curr] == 0){
                int temp = nums[left];
                nums[left++] = nums[curr];
                nums[curr++] = temp;
            }
            else if(nums[curr] == 2){
                int temp = nums[right];
                nums[right--] = nums[curr];
                nums[curr] = temp;  
            }
            else{
                curr++;
            }
        }

    }
}

// l = 0
// r = 6

// [2,0,2,1,1,0]
//  l         r

// [0 0 1 1 2 2]
//      l    
//        r
           
// swap 2 0


